jerry-core/jerry.c

```diff
@@ -1845,81 +1828,112 @@ snapshot_add_compiled_code (ecma_compiled_code_t *compiled_code_p, /**< compiled
+  ecma_compiled_code_t *copied_compiled_code_p;
 
-  compiled_code_map_entry_t *new_entry;
-  new_entry = (compiled_code_map_entry_t *) mem_heap_alloc_block (sizeof (compiled_code_map_entry_t));
+  copied_compiled_code_p = (ecma_compiled_code_t *) (snapshot_buffer_p + snapshot_buffer_write_offset);
 
-  if (new_entry == NULL)
+  if (!(compiled_code_p->status_flags & CBC_CODE_FLAGS_FUNCTION))
   {
-    snapshot_error_occured = true;
-    return;
-  }
+    /* Regular expression. */
+    if (snapshot_buffer_write_offset + sizeof (ecma_compiled_code_t) > snapshot_buffer_size)
+    {
+      snapshot_error_occured = true;
+      return 0;
+    }
 
-  ECMA_SET_POINTER (new_entry->next_cp, snapshot_map_entries_p);
-  ECMA_SET_POINTER (new_entry->compiled_code_cp, compiled_code_p);
+    snapshot_buffer_write_offset += sizeof (ecma_compiled_code_t);
 
-  new_entry->offset = (uint16_t) (snapshot_buffer_write_offset >> MEM_ALIGNMENT_LOG);
-  snapshot_map_entries_p = new_entry;
+    mem_cpointer_t pattern_cp = ((re_compiled_code_t *) compiled_code_p)->pattern_cp;
+    ecma_string_t *pattern_string_p = ECMA_GET_NON_NULL_POINTER (ecma_string_t,
+                                                                 pattern_cp);
 
-  const void *data_p = (const void *) compiled_code_p;
+    ecma_length_t pattern_size = ecma_string_get_size (pattern_string_p);
 
-  if (!(compiled_code_p->status_flags & CBC_CODE_FLAGS_FUNCTION))
-  {
-    size += (uint32_t) sizeof (uint16_t);
-  }
-  else
-  {
-    JERRY_ASSERT (regexp_pattern == NULL);
-  }
+    MEM_DEFINE_LOCAL_ARRAY (buffer_p, pattern_size, lit_utf8_byte_t);
 
-  if (!jrt_write_to_buffer_by_offset (snapshot_buffer_p,
-                                      snapshot_buffer_size,
-                                      &snapshot_buffer_write_offset,
-                                      &size,
-                                      sizeof (uint32_t)))
-  {
-    snapshot_error_occured = true;
-    return;
-  }
-
-  if (!(compiled_code_p->status_flags & CBC_CODE_FLAGS_FUNCTION))
-  {
-    size -= (uint32_t) sizeof (uint16_t);
+    ecma_string_to_utf8_string (pattern_string_p, buffer_p, pattern_size);
 
     if (!jrt_write_to_buffer_by_offset (snapshot_buffer_p,
                                         snapshot_buffer_size,
                                         &snapshot_buffer_write_offset,
-                                        &compiled_code_p->status_flags,
-                                        sizeof (uint16_t)))
+                                        buffer_p,
+                                        pattern_size))
     {
       snapshot_error_occured = true;
-      return;
     }
 
-    data_p = regexp_pattern;
+    MEM_FINALIZE_LOCAL_ARRAY (buffer_p);
+
+    snapshot_buffer_write_offset = JERRY_ALIGNUP (snapshot_buffer_write_offset, MEM_ALIGNMENT);
+
+    /* Regexp character size is stored in refs. */
+    copied_compiled_code_p->refs = (uint16_t) pattern_size;
```

Comment:
This seems to rely on the re_compiled_code_t and ecma_compiled_code_t structs to stay stable relative to each other. Why not create a new re_compiled_code_t variable pointing at the same memory instead?